File: proposed-projects/round-trip-experiment/reverse/20260909T031655Z_89df6247/20260910T063135Z_0babbdab/ported_implementation/src/utils/errors/grammarParseError.ts
import { buildErrorPosition } from './buildErrorPosition.js';

export const GRAMMAR_PARSER_ERROR_HEADER_MESSAGE = (reason: string): string =>
  `Failed to parse grammar: ${reason}`;

export class GrammarParseError extends Error {
  public readonly name = 'GrammarParseError';
  public readonly grammar: string;
  public readonly reason: string;
  public readonly pos: number;

  public constructor(grammar: string, pos: number, reason: string) {
    super(
      [
        GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
        '',
        ...buildErrorPosition(grammar, pos),
      ].join('\n')
    );
    Object.setPrototypeOf(this, GrammarParseError.prototype);
    this.grammar = grammar;
    this.reason = reason;
    this.pos = pos;
  }
}

File: proposed-projects/round-trip-experiment/reverse/20260909T123239Z_5982b634/20260910T070936Z_fb630842/ported_implementation/gbnf/utils/errors/grammar_parse_error.py
from __future__ import annotations

from .build_error_position import build_error_position


def GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason: str) -> str:
    return f'Failed to parse grammar: {reason}'


class GrammarParseError(Exception):
    def __init__(self, grammar: str, pos: int, reason: str):
        super().__init__(
            '\n'.join(
                [
                    GRAMMAR_PARSER_ERROR_HEADER_MESSAGE(reason),
                    '',
                    *build_error_position(grammar, pos),
                ]
            )
        )
        self.name = 'GrammarParseError'
        self.grammar = grammar
        self.pos = pos
        self.reason = reason

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, GrammarParseError):
            return NotImplemented
        return (
            self.grammar == other.grammar
            and self.pos == other.pos
            and self.reason == other.reason
        )

    def __hash__(self) -> int:
        return hash((self.grammar, self.pos, self.reason))
